pkg/pool-weighted/contracts/smart/IndexPool.sol

```diff
@@ -209,15 +233,100 @@ contract IndexPool is BaseWeightedPool, ReentrancyGuard {
 
         (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
 
-        _startGradualWeightChange(startTime, endTime, _getNormalizedWeights(), endWeights, tokens);
+        _startGradualWeightChange(
+            startTime,
+            endTime,
+            _getNormalizedWeights(),
+            endWeights,
+            tokens,
+            new uint256[](tokens.length)
+        );
     }
 
-    function reweighTokens(address[] calldata tokens, uint256[] calldata desiredWeights) public {
+    function reweighTokens(IERC20[] calldata tokens, uint256[] calldata desiredWeights) public {
         uint256 endTime = _getMiscData().decodeUint32(_END_TIME_OFFSET);
         require(block.timestamp >= endTime, "Weight change is already in process");
+        InputHelpers.ensureInputLengthMatch(tokens.length, desiredWeights.length);
+        uint256 changeTime = _calcReweighTime(tokens, desiredWeights);
+        _updateWeightsGradually(block.timestamp, block.timestamp.add(changeTime), desiredWeights);
+    }
+
+    function reindexTokens(
+        IERC20[] memory tokens,
+        uint256[] memory desiredWeights,
+        uint256[] memory minimumBalances
+    ) external {
+        InputHelpers.ensureInputLengthMatch(tokens.length, desiredWeights.length, minimumBalances.length);
+
+        /*
+            assemble params for IndexPoolUtils._normalizeInterpolated:
+        */
+        // the initial weights in the pool (here a new token has a weight of zero)
+        uint256[] memory baseWeights = new uint256[](tokens.length);
+        // the weights that are fixed and that the other tokens need to be adjusted by
+        uint256[] memory fixedWeights = new uint256[](tokens.length);
+        // we need to store the final desired weight of a new tokensince initially it will be set to 1%
+        uint256[] memory newTokenTargetWeights = new uint256[](tokens.length);
+
+        /*
+            this is some mambojambo to get an array that only contains the
+            addresses of the new tokens
+        */
+        IERC20[] memory newTokensContainer = new IERC20[](tokens.length);
+        uint8 newTokenCounter = 0;
+
+        for (uint8 i = 0; i < tokens.length; i++) {
+            require(minimumBalances[i] != 0, "Invalid zero minimum balance");
+            bytes32 currentTokenState = _tokenState[IERC20(tokens[i])];
+
+            // check if token is new token by checking if no startTime is set
+            if (currentTokenState.decodeUint64(_START_WEIGHT_OFFSET) == 0) {
```

Comment:
Would not the whole `currentTokenState` variable be empty when the token is new. As in this: `_tokenState[IERC20(tokens[i])]` would just be `_tokenState[tokenThatIsNotInMapping]`?

Reply:
This might be more for my understanding then that something had to change

Reply:
Yes but how would you test for the whole `currentTokenState` being empty? I initially checked if `currentTokenState` == 0, but choosing one specific value that supposedly should be set for a new token (the `startWeight`) seemed safer to me.